File: head_to_offer/34/hto_34.py
from typing import List, Dict
from queue import Queue
from collections import deque
import math
import itertools


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
    
class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        if not root: 
            return []
        res, tmp = [], []
        self.dfs(root, sum, res, tmp)
        return res
        
    def dfs(self, node: TreeNode, num: int, res: List, tmp: List) -> None:
        if not node:
            return
        tmp.append(node.val)
        if not node.left and not node.right and num == node.val:
            res.append(tmp[:])
            tmp.pop()
            return
        self.dfs(node.left,  num - node.val, res, tmp)
        self.dfs(node.right, num - node.val, res, tmp)
        tmp.pop()



if __name__ == "__main__":
    node1 = TreeNode(5)
    node2 = TreeNode(4)
    node3 = TreeNode(8)
    node4 = TreeNode(11)
    node5 = TreeNode(13)
    node6 = TreeNode(4)
    node7 = TreeNode(7)
    node8 = TreeNode(2)
    node9 = TreeNode(5)
    node10 = TreeNode(1)
    node1.left = node2
    node1.right = node3
    node2.left = node4
    node3.left = node5
    node3.right = node6
    node4.left = node7
    node4.right = node8
    node6.left = node9
    node6.right = node10
    s = Solution()
    print(s.pathSum(node1, 22))

File: hto/34/hto_34.cpp
#include <iostream>
#include <string>
#include <vector>
#include <climits>
#include <algorithm>
#include <cstring>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <stack>
#include <bitset>
#include <array>
#include <queue>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x): val(x), left(NULL), right(NULL) { }
};

class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        if (!root) return {};
        vector<vector<int>> res;
        stack<pair<TreeNode*, vector<int>>> stack;
        stack.push({root, {root->val}});
        while (!stack.empty()) {
            pair<TreeNode*, vector<int>> p = stack.top();
            stack.pop();
            int sumVector = accumulate(p.second.begin(), p.second.end(), 0);
            if (!p.first->left && !p.first->right && sumVector == sum)
                res.emplace_back(p.second);
            if (p.first->left) {
                vector<int> tmp(p.second);
                tmp.emplace_back(p.first->left->val);
                stack.push({p.first->left, tmp});
            }
            if (p.first->right) {
                vector<int> tmp(p.second);
                tmp.emplace_back(p.first->right->val);
                stack.push({p.first->right, tmp});
            }
        }
        return res;
    }
};

int main() {
    TreeNode *node1 = new TreeNode(3);
    TreeNode *node2 = new TreeNode(9);
    TreeNode *node3 = new TreeNode(20);
    TreeNode *node4 = new TreeNode(15);
    TreeNode *node5 = new TreeNode(7);
    node1->left = node2; node1->right = node3;
    node2->left = node4; node3->right = node5;
    Solution s;
    for (vector<int> v : s.pathSum(node1, 27)) {
        for (int i : v) {
            cout << i << " ";
        }
        cout << "\n";
    }
}

